downstairs/src/region.rs

```diff
@@ -1270,6 +1346,75 @@ impl Extent {
 
         Ok(())
     }
+
+    /// Rehash the entire file. Remove any stored hash/encryption contexts
+    /// that do not correlate to data currently stored on disk. This is
+    /// primarily when opening an extent after recovering from a crash, since
+    /// irrelevant hashes will normally be cleared out during a flush().
+    ///
+    /// By default this function will only do work if the extent is marked
+    /// dirty. Set `force_override_dirty` to `true` to override this behavior.
+    /// This override should generally not be necessary, as the dirty flag
+    /// is set before any contexts are written.
+    #[instrument]
+    fn fully_rehash_and_clean_all_stale_contexts(
+        &self,
+        force_override_dirty: bool,
+    ) -> Result<(), CrucibleError> {
+        let mut inner = self.inner();
+
+        if !force_override_dirty && !inner.dirty()? {
+            return Ok(());
+        }
+
+        // Just in case, let's be very sure that the file on disk is what it should be
+        if let Err(e) = inner.file.sync_all() {
+            crucible_bail!(
+                IoError,
+                "extent {}: fsync 1 failure during full rehash: {:?}",
+                self.number,
+                e
+            );
+        }
+
+        inner.file.seek(SeekFrom::Start(0))?;
+
+        // Buffer the file so we dont spend all day waiting on syscall
+        let mut inner_file_buffered =
+            BufReader::with_capacity(64 * 1024, &inner.file);
+
+        // This gets filled one block at a time for hashing
+        let mut block = vec![0; self.block_size as usize];
+
+        // The vec of hashes that we'll pass off to truncate...()
+        let mut extent_block_indexes_and_hashes =
+            Vec::with_capacity(self.extent_size.value as usize);
+
+        // Stream the contents of the file and rehash them.
+        for i in 0..self.extent_size.value as usize {
+            inner_file_buffered.read_exact(&mut block)?;
+            extent_block_indexes_and_hashes
+                .push((i, integrity_hash(&[&block])));
+        }
+
+        // NOTE on safety: Unlike BufWriter, BufReader drop() does not have
+        // side-effects. There are no unhandled errors here.
+        drop(inner_file_buffered);
+
+        inner.truncate_encryption_contexts_and_hashes(
+            extent_block_indexes_and_hashes,
+        )?;
+
+        inner.file.seek(SeekFrom::Start(0))?;
+
+        // XXX should we be clearing the dirty flag here? If we do, should we
```

Comment:
No clearing of dirty bit here, we want the repair to know that this block was dirty.